avl_tree/avl.py

```diff
@@ -0,0 +1,146 @@
+            self.search(root.left,value)
+            self.search(root.right,value)
+        elif root.left!=None:
+            self.search(root.left,value)
+        elif root.right!=None:
+            self.search(root.right,value)
+
+    # def height(self,root): #root is a node
+    #     if root.left!=None and root.right!=None:
+    #         return 1+ max(self.height(root.left),self.height(root.right))
+    #     elif root.left!=None:
+    #         return 1+ self.height(root.left)
+    #     elif root.right!=None:
+    #         return 1+ self.height(root.right)
+    #     else:
+    #         return 1
+    #
+    # def getheight(self,value):
+    #     self.search(self.head,value)
+
+    # def isroot(self,node):
+    #     if node.data==self.head:
+    #         return True
+    #     else:
+    #         return False
+
+    def rotateleft(self,rotroot):
+        new=rotroot.right
+        rotroot.right=new.left
+        if new.left!=None:
+            new.left.parent=rotroot
+        new.parent=rotroot.parent
+        if rotroot.parent==None:
+            self.head=new
+        else:
+            if rotroot.left!=None:
+                rotroot.parent.left=new
+            else:
+                rotroot.parent.right=new
+        new.left=rotroot
+        rotroot.parent=new
+
+    def rotateright(self,rotroot):
+        new=rotroot.left
+        rotroot.left=new.right
+        if new.right!=None:
+            new.right.parent=rotroot
+        new.parent=rotroot.parent
+        if rotroot.parent==None:
+            self.head=new
+        else:
+            if rotroot.right!=None:
+                rotroot.parent.right=new
+            else:
+                rotroot.parent.left=new
+        new.right=rotroot
+        rotroot.parent=new
+
+    def printme(self,root):
+        if root!=None:
+            print (root.data)
+            self.printme(root.left)
+            self.printme(root.right)
+
+    def printtree(self):
+        self.printme(self.head)
+
+    # def rrotate(self,value):
+    #     n=self.search(self.head,value)
+    #     self.rotateright(n)
+    #
+    # def lrotate(self,value):
+    #     n=self.search(self.head,value)
+    #     self.rotateleft(n)
+
+    def rebalance(self,root):
+        if root.balance() <0:
+            if root.right.balance > 0:
+                self.rotateright(root.right)
+                self.rotateleft(root)
+            else:
+                self.rotateleft(root)
+        elif root.balance() >0:
+            if root.left.balance() <0:
+                self.rotateleft(node.left)
+                self.rotateright(node)
+            else:
+                self.rotateright(root)
+
+    def updatebal(self,root):
+        temp=root
+        while temp!=None:
+            if temp!=None:
+                if temp.balance() > 1 or temp.balance < -1:
+                    self.rebalance(temp)
+            temp=temp.parent
+
+
+a=tree(5)
+a.add(7)
+a.add(6)
+#a.printtree()
+#a.lrotate(5)
+#a.printtree()
```

Comment:
Line contains following spacing inconsistencies:
- No newline at EOF.

*Origin: SpaceConsistencyBear, Section: `python spacing`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpswghsfzh/avl_tree/avl.py
+++ b/tmp/tmpswghsfzh/avl_tree/avl.py
@@ -143,4 +143,4 @@
 a.add(6)
 #a.printtree()
 #a.lrotate(5)
-#a.printtree()+#a.printtree()
```